### DISCUSSION TITLE
Sending Bulk Email

### DISCUSSION THREAD

**Discussion by ljfreelancer88** [wneessen/go-mail#73] [Q&A] [answered]
Hi, I'm planning to use `go-mail` in sending bulk email but I can't find an example in the documentation. The code below is what I have in mind, it's basically wrapping with `for` loop. Any correction on the code is appreaciated. Thank you in advance.

```go
m := mail.NewMsg()
for _, v := range mail {
	if err := m.FromFormat("Toni Tester", "sender@example.com"); err != nil {
		fmt.Printf("failed to set FROM address: %s\n", err)
		os.Exit(1)
	}
	if err := m.To(`"v.Name <v.Email>"`); err != nil {
		fmt.Printf("failed to set TO address: %s\n", err)
		os.Exit(1)
	}
	m.CcIgnoreInvalid("cc@example.com", "invalidaddress+example.com")

	m.Subject("This is a great email")
	m.SetDate()
	m.SetMessageID()
	m.SetBulk()
	m.SetImportance(mail.ImportanceHigh)
	m.SetBodyString(mail.TypeTextPlain, "This is a great message body text.")
	m.AttachFile("/home/ttester/test.txt", mail.WithFileName("attachment.txt"))

	host := "relay.example.com"
	c, err := mail.NewClient(host,
		mail.WithSMTPAuth(mail.SMTPAuthPlain), mail.WithUsername("ttester"),
		mail.WithPassword("V3rySecUr3!Pw."), mail.WithTLSPolicy(mail.TLSMandatory))
	if err != nil {
		fmt.Printf("failed to create new mail client: %s\n", err)
		os.Exit(1)
	}

	if err := c.DialAndSend(m); err != nil {
		fmt.Printf("failed to send mail: %s\n", err)
		os.Exit(1)
	}

	fmt.Println("Mail successfully sent.")
}
```

**Comment by wneessen** [comment, marked as answer]
Hi @ljfreelancer88,

that's actually a good point. We currently don't have a proper bulk-mail example. I'll add one to the documentation in the next days. For now, a few changes I would suggest for your code:

* Instead of using `m.To(`"v.Name <v.Email>"`);` you can use `m.AddToFormat(v.Name, v.Email)`which will take care of the proper formating for you.
* In your example you create a new client in the for loop. This is not needed. go-mail can reuse the client connection and takes a list of `*mail.Msg` in the `c.DialAndSend()` call. Therefore I suggest to create an empty mail message slice `[]*mail.Msg` and only append the prepared mails into that in the for loop and eventually provide the message slice to the `c.DialAndSend()` method
* Generally the example code you've taken has lots of feature examples that you might not need. I. e. I would remove the `m.SetImportance(mail.ImportanceHigh)` for a bulk mailing unless it's really of high importance.

I've composed an example code that should fit your needs better below.

Hope this helps!
Winni


```go
package main

import (
	"fmt"
	"github.com/wneessen/go-mail"
	"os"
)

func main() {
	type user struct {
		Name  string
		Email string
	}
	userList := []user{
		{"Toni Tester", "toni.tester@example.com"},
		{"Tina Tester", "tina.tester@example.com"},
		{"John Doe", "john.doe@example.com"},
	}

	var mailList []*mail.Msg
	for _, v := range userList {
		m := mail.NewMsg()
		if err := m.FromFormat("Stephen Sender", "sender@example.com"); err != nil {
			fmt.Printf("failed to set FROM address: %s\n", err)
			os.Exit(1)
		}
		if err := m.AddToFormat(v.Name, v.Email); err != nil {
			fmt.Printf("failed to set TO address: %s\n", err)
			os.Exit(1)
		}
		m.Subject("This is a great email")
		m.SetDate()
		m.SetMessageID()
		m.SetBodyString(mail.TypeTextPlain, "This is a great message body text.")
		mailList = append(mailList, m)
	}

	host := "relay.example.com"
	c, err := mail.NewClient(host,
		mail.WithSMTPAuth(mail.SMTPAuthPlain), mail.WithUsername("ttester"),
		mail.WithPassword("V3rySecUr3!Pw."), mail.WithTLSPolicy(mail.TLSMandatory))
	if err != nil {
		fmt.Printf("failed to create new mail client: %s\n", err)
		os.Exit(1)
	}

	if err := c.DialAndSend(mailList...); err != nil {
		fmt.Printf("failed to send mail: %s\n", err)
		os.Exit(1)
	}
	fmt.Printf("%d mails successfully sent.", len(mailList))
}
```

**Reply by wneessen** [reply]
Hi again @ljfreelancer88,

I've taken some time today, to build a proper documented bulk mailer example and added it to our documentation website. You can find it here: https://go-mail.dev/examples/bulk-mailer/

It's pretty similar to what I already posted here, but is a bit more detailed and describes all of the different steps taken.

Hope this helps! Please let me know if you have any comments or things that should be improved.

**Reply by ljfreelancer88** [reply]
Thank You so much! Can we add both html and plain text for the template and add a little bit of html tag to the html template, to make it clear where to put the all html tag, since there are two template involve. I'm looking at the "newbie" perspective. The code below is just a sample. 

Regarding parsing these templates, I'll leave it all to you :D 

```golang
const htmlBodyTemplate = `<html>Hi {{.Firstname}},

we are writing your to let you know that this week we have an amazing offer for you.
Using the coupon code <em>"GOMAIL"</em> you will get a <strong>20% discount</strong> on all our products in our
online shop.

Check out our latest offer on https://acme.com and use your discount code today!

Your marketing team
  at ACME Inc. </html>`
  
const plainTextBodyTemplate = `Hi {{.Firstname}},

we are writing your to let you know that this week we have an amazing offer for you.
Using the coupon code "GOMAIL" you will get a 20% discount on all our products in our
online shop.

Check out our latest offer on https://acme.com and use your discount code today!

Your marketing team
  at ACME Inc.`
```

**Reply by wneessen** [reply]
Good point. I'll get the HTML into the example as well. I'll let you know once that has been updated

**Reply by wneessen** [reply]
https://go-mail.dev/examples/bulk-mailer/ now covers both, text- and HTML templating

**Reply by ljfreelancer88** [reply]
I got `failed to deliver mail: dial failed: dial tcp 18.215.44.90:25: i/o timeout` at first because I didn't set the port `2525`. Though it was in the [introduction](https://go-mail.dev/getting-started/introduction/). Maybe we can add that too in the examples for consitency. Thank you.

**Reply by wneessen** [reply]
Port 2525 is a non-standard port. Therefore I would not got that route. Most SMTP services work on port 25 or 587.
